### ISSUE TITLE
Add normalized equivalent of YieldProcessor, retune some spin loops 

### ISSUE THREAD

**Issue by benaadams** [mono/mono#9011]
For the Skylake architecture and above Intel decided to change the `PAUSE` instruction from 10 cycles to 140 cycles.

This causes upto a x14 degradation in `YieldProcessor` type spinning (contended `lock`, `ManualResetEventSlim`, use of `Thread.SpinWait`, etc)

See issue in https://github.com/dotnet/coreclr/issues/13388
Also CoreClr fix in https://github.com/dotnet/coreclr/pull/13670

The TechEmpower hardware in the physical scenerios uses Skylake architecture, so it would be good to eliminate this as an difference (also is good to fix for people on Skylake and above)

/cc @migueldeicaza

**Comment by kumpera** [comment]
We don't play those coreclr shenanigans with pause, so I fail to see where we'd need that.

**Comment by lewurm** [comment]
we don't do spinning using `pause` in the runtime.

**Comment by kumpera** [comment]
More importantly, we don't plan on having the sort of aggressive spinning that coreclr does as it comes with at the expense of power usage - which we care more than the extra few % of perf in some workloads.

**Comment by benaadams** [comment]
> More importantly, we don't plan on having the sort of aggressive spinning that coreclr does as it comes at the expense of power usage 

🤔 

https://github.com/mono/mono/blob/bf3c71ef3f9e160951d8778e16cabe547e4360d1/mcs/class/referencesource/mscorlib/system/threading/SpinLock.cs#L378-L385

https://github.com/mono/mono/blob/bf3c71ef3f9e160951d8778e16cabe547e4360d1/mcs/class/referencesource/mscorlib/system/threading/ManualResetEventSlim.cs#L601-L603

**Comment by benaadams** [comment]
From the coreclr issue

> I have been tracking an unusual MAJOR regression (as in capital letters major) from 60K req/sec (running the loading and the target process at the same machine) to 3K req/sec that may be related to this. We have only seen this after upgrading the solution to CoreCLR 2.0. Under profiler the only suspect for the odd behavior is the **ManualResetEventSlim**. We see an inordinate amount of CPU time being assigned when that should have been very low.

Though it looks like `SpinWait` translates to `NOP` (1 cycle) rather than `PAUSE` (10 cycles -> 140 cycles) in Mono; which is more power consuming and a single 1 cycle spin is unlikely to resolve any contention so it will spin in a very tight loop. 

Using `NOP` is still a tight loop burning CPU, where as `PAUSE` releases all CPU resource to the second thread under hyperthreading or moves the CPU to a low power state.

From [Benefitting Power and Performance Sleep Loops - Intel](https://software.intel.com/en-us/articles/benefitting-power-and-performance-sleep-loops)

> Another issue is that immediately giving up a time slice means the running thread will be context switched out. A context switch costs something on the order of 5,000 cycles, so getting switched out and switched back in **means the processor has wasted at least 10,000 cycles of overhead**, which is not helping the workload get completed any faster. Very often, these loops lead to very high context switch rates, which are signs of overhead and possible opportunities for performance gains.
>
> Fortunately, you have some options for help mitigating the overhead, saving power, and getting a nice boost in performance.
>
> ### Spinning Out of Control
>
> If you are using a threading library, you may not have control over the spin algorithms implemented.  During performance analysis, you may see a high volume of context switches, calls to Sleep or SwitchToThread, and high processor utilization tagged to the threading library.  In these situations, it is worth looking at alternative threading libraries to determine if their spin algorithms are more efficient.
>
> ### Resolving the Problems
>
> The approach we recommend in such an algorithm is akin to a more gradual back off. First, we allow the thread to spin on the lock for a brief period of time, but instead of fully spinning, we use the pause instruction in the loop. Introduced with the Intel® Streaming SIMD Extensions 2 (Intel® SSE2) instruction set, the pause instruction gives a hint to the processor that the calling thread is in a "spin-wait" loop. In addition, the pause instruction is a no-op when used on x86 architectures that do not support Intel SSE2, meaning it will still execute without doing anything or raising a fault. While this means older x86 architectures that don’t support Intel SSE2 won’t see the benefits of the pause, it also means that you can keep one straightforward code path that works across the board.
>
> Essentially, the pause instruction delays the next instruction's execution for a finite period of time. By delaying the execution of the next instruction, the processor is not under demand, and parts of the pipeline are no longer being used, which in turn reduces the power consumed by the processor.
>
> The pause  instruction can be used in conjunction with a Sleep(0) to construct something similar to an exponential back-off in situations where the lock or more work may become available in a short period of time, and the performance may benefit from a short spin in ring 3.

**Comment by benaadams** [comment]
And depending how much framework or coreclr code is consumed: 
![image](https://user-images.githubusercontent.com/1142958/41549824-ee1dbdf6-731e-11e8-9ced-9bdf22d6914f.png)
From [Why Skylake CPUs Are Sometimes 50% Slower – How Intel Has Broken Existing Code](https://aloiskraus.wordpress.com/2018/06/16/why-skylakex-cpus-are-sometimes-50-slower-how-intel-has-broken-existing-code/)

**Comment by kumpera** [comment]
Fair enough. We should add support for PAUSE / YIELD and tune those BCL knobs to be geared towards low power.

**Comment by kumpera** [comment]
Since we're here.
We should audit usage of Environment.ProcessorCount and PlatformHelper.ProcessorCount.
BCL code should have a very good reason to use former (including caching its value) and not the later.

**Comment by benaadams** [comment]
> We should audit usage of Environment.ProcessorCount and PlatformHelper.ProcessorCount.
BCL code should have a very good reason to use former (including caching its value) and not the later.

I believe the caching is because checking the cpu count is an expensive synchronising request; and cpus counts don't change that frequently? (so it only rechecks every N requests)

The risk of not caching is it becomes expensive; so users then manually cache it in a `readonly static` ending up with a fixed value rather than one which at least updates?

**Comment by pr0vieh** [comment]
@benaadams ❤️ 
I could not explain why after a server / cpu change from Ivy to Skylake mono performance has dropped
now it's clear!

**Comment by jaykrell** [comment]
Doesn’t spinning save battery as often as it is faster on non-battery? I mean, I understand it is all heuristic and that spinning is worse under contention.

In either case, ideally, this belongs at the next layer down, in usermode and not in the kernel, and on Unix we probably already get that, possibly all systems.

i.e. Linux kernel only has futex and pthreads probably only uses that after a spin.
All the pthread systems are probably similar in this regard.

Win32 SRWLock, CondVar, CritSec all spin. They are all usermode constructs.

Win32 Event, Mutex, Semaphore is what you want to avoid using directly, for performance, unless you have highly contended locks, or cross-process locks, or named locks, all of which should be rare. They are kernel constructs.

**Comment by kumpera** [comment]
Hi @jaykrell, look at the links provided by @benaadams to better contextualize your comments.

**Comment by kumpera** [comment]
@benaadams while doing a basic review of the BCL I found a couple places that look maybe fishy (perf or correctness-wise).

The issue with using, and caching, Environment.ProcessorCount is that CPU could do change dynamically on mobile and we need to react to that otherwise perf suffers significantly.

**Comment by benaadams** [comment]
> The issue with using, and caching, Environment.ProcessorCount is that CPU could do change dynamically on mobile and we need to react to that otherwise perf suffers significantly.

Its definitely a balance; something that checks the count very infrequently isn't going to adapt anyway; and something that checks it hundreds of times per second will be checking it far faster than the mobile would change CPU count?

**Comment by benaadams** [comment]
It checking elapsed ticks more or less costly than checking CPU count? (i.e cache for x time rather than request count)

**Comment by kumpera** [comment]
It can be a couple of orders of magnitude faster on some targets.

**Comment by jaykrell** [comment]
How does CPU count change? I understand hot-add?remove?swap of CPUs on high end servers.
On Mobile? In VMs? Is there a notification one can get? Granted, nice to have one code path that handles high end servers and mobile, so CPU count can change in general.
Checking tickcount on Windows is super fast, other than call/ret, it is just reading a fixed address.
Hopefully other systems similar.
I'm not sure how checking CPU count works.

**Comment by benaadams** [comment]
Arm uses [big.LITTLE](https://en.wikipedia.org/wiki/ARM_big.LITTLE) where they have high power cores and low power cores; they use the high power cores when CPU load is high, then shift the loads to the low power cores when the CPU load is lower so the number of CPU cores available increases and decreases based on load (or battery power remaining)

**Comment by benaadams** [comment]
Sorry, big.LITTLE is match sets; is [DynamicIQ](https://developer.arm.com/technologies/dynamiq) that changes layouts

**Comment by jaykrell** [comment]
DynamicIQ is powering on/off the little cores, effectively adding/removing them?
The kernel will see and expose it like that? Wow.
